pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,212 @@
+package alertmanager
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/alertmanager/cluster"
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// state represents the Alertmanager silences and notification log internal state
+type state struct {
+	userID string
+
+	mtx    sync.Mutex
+	states map[string]cluster.State
+	logger log.Logger
+	reg    prometheus.Registerer
+
+	replicationFactor  int
+	replicateStateFunc func(context.Context, string, *clusterpb.Part) error
+
+	partialStateMergesTotal  *prometheus.CounterVec
+	partialStateMergesFailed *prometheus.CounterVec
+	stateReplicationTotal    *prometheus.CounterVec
+	stateReplicationFailed   *prometheus.CounterVec
+
+	msgc   chan *clusterpb.Part
+	stopc  chan struct{}
+	readyc chan struct{}
+}
+
+func (s *state) Name() string {
+	return ""
+}
+
+func (s *state) Address() string {
+	return ""
+}
+
+func (s *state) Peers() []cluster.Node {
+	return nil
+}
+
+// newReplicatedStates creates a new state struct, which manages state to be replicated between alertmanagers.
+func newReplicatedStates(userID string, rf int, f func(context.Context, string, *clusterpb.Part) error, stopc chan struct{}, l log.Logger, r prometheus.Registerer) *state {
```

Comment:
Design wise, I'm not a big fan of taking in input `stopc`. I think `state` should self-manage its lifecycler (eg. could be a service).

Reply:
Alertmanager isn't so it seemed like overkill, happy to do it if you think otherwise.